database/raft/state/state.go

```diff
@@ -0,0 +1,223 @@
+package state
+
+import (
+	"bytes"
+	"context"
+
+	"github.com/golang/protobuf/proto"
+
+	"chain/database/raft/internal/statepb"
+	"chain/errors"
+	"chain/log"
+)
+
+const nextNodeID = "raft/nextNodeID"
+
+var ErrAlreadyApplied = errors.New("entry already applied")
+
+// State is a general-purpose data store designed to accumulate
+// and apply replicated updates from a raft log.
+type State struct {
+	state        map[string][]byte
+	peers        map[uint64]string // id -> addr
+	appliedIndex uint64
+	version      map[string]uint64 //key -> value index
+}
+
+// New returns a new State
+func New() *State {
+	return &State{
+		state:   map[string][]byte{nextNodeID: []byte("2")},
+		peers:   make(map[uint64]string),
+		version: make(map[string]uint64),
+	}
+}
+
+// SetPeerAddr sets the address for the given peer.
+func (s *State) SetPeerAddr(id uint64, addr string) {
+	s.peers[id] = addr
+}
+
+// GetPeerAddr gets the current address for the given peer, if set.
+func (s *State) GetPeerAddr(id uint64) (addr string) {
+	return s.peers[id]
+}
+
+// RemovePeerAddr deletes the current address for the given peer if it exists.
+func (s *State) RemovePeerAddr(id uint64) {
+	delete(s.peers, id)
+}
+
+// RestoreSnapshot decodes data and overwrites the contents of s.
+// It should be called with the retrieved snapshot
+// when bootstrapping a new node from an existing cluster
+// or when recovering from a file on disk.
+func (s *State) RestoreSnapshot(data []byte, index uint64) error {
+	s.appliedIndex = index
+	//TODO (ameets): think about having statepb in state for restore
+	snapshot := &statepb.Snapshot{}
+	err := proto.Unmarshal(data, snapshot)
+	s.peers = snapshot.Peers
+	s.state = snapshot.State //TODO (ameets): need to add version here
```

Comment:
Yes. My next three follow up issues are:

1. Evicting nodes from the cluster when they go offline
2. Protecting against incompatible versions
3. Forcing corectl to use the API.